### ISSUE TITLE
How to Configure the Number of Layers and Nodes in a Neural Network

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#445]
How to Configure the Number of Layers and Nodes in a Neural Network<br>
<br>
https://ift.tt/2LO9pHE<br>
<br>
Jason Brownlee<br>
<br>
<p>Artificial neural networks have two main hyperparameters that control the architecture or topology of the network: the number of layers and the number of nodes in each hidden layer.</p>
<p>You must specify values for these parameters when configuring your network.</p>
<p>The most reliable way to configure these hyperparameters for your specific predictive modeling problem is via systematic experimentation with a robust test harness.</p>
<p>This can be a tough pill to swallow for beginners to the field of machine learning, looking for an analytical way to calculate the optimal number of layers and nodes, or easy rules of thumb to follow.</p>
<p>In this post, you will discover the roles of layers and nodes and how to approach the configuration of a multilayer perceptron neural network for your predictive modeling problem.</p>
<p>After reading this post, you will know:</p>
<ul>
<li>The difference between single-layer and multiple-layer perceptron networks.</li>
<li>The value of having one and more than one hidden layers in a network.</li>
<li>Five approaches for configuring the number of layers and nodes in a network.</li>
</ul>
<p>Let&rsquo;s get started.</p>
<div class="wp-caption">
<img src="https://machinelearningmastery.com/wp-content/uploads/2018/07/How-to-Configure-the-Number-of-Layers-and-Nodes-in-a-Neural-Network.jpg" alt="How to Configure the Number of Layers and Nodes in a Neural Network" width="640" class="wp-image-5859" height="449">
<p class="wp-caption-text">How to Configure the Number of Layers and Nodes in a Neural Network<br>
Photo by <a href="https://www.flickr.com/photos/scorsese/7272760288/">Ryan</a>, some rights reserved.</p>
</div>
<h2>Overview</h2>
<p>This post is divided into four sections; they are:</p>
<ol>
<li>The Multilayer Perceptron</li>
<li>How to Count Layers?</li>
<li>Why Have Multiple Layers?</li>
<li>How Many Layers and Nodes to Use?</li>
</ol>
<h2>The Multilayer Perceptron</h2>
<p>A node, also called a neuron or Perceptron, is a computational unit that has one or more weighted input connections, a transfer function that combines the inputs in some way, and an output connection.</p>
<p>Nodes are then organized into layers to comprise a network.</p>
<p>A single-layer artificial neural network, also called a single-layer, has a single layer of nodes, as its name suggests. Each node in the single layer connects directly to an input variable and contributes to an output variable.</p>
<blockquote>
<p>Single-layer networks have just one layer of active units. Inputs connect directly to the outputs through a single layer of weights. The outputs do not interact, so a network with N outputs can be treated as N separate single-output networks.</p>
</blockquote>
<p>&mdash; Page 15, <a href="https://amzn.to/2vhyW8j">Neural Smithing: Supervised Learning in Feedforward Artificial Neural Networks</a>, 1999.</p>
<p>A single-layer network can be extended to a multiple-layer network, referred to as a Multilayer Perceptron. A Multilayer Perceptron, or MLP for sort, is an artificial neural network with more than a single layer.</p>
<p>It has an input layer that connects to the input variables, one or more hidden layers, and an output layer that produces the output variables.</p>
<blockquote>
<p>The standard multilayer perceptron (MLP) is a cascade of single-layer perceptrons. There is a layer of input nodes, a layer of output nodes, and one or more intermediate layers. The interior layers are sometimes called &ldquo;hidden layers&rdquo; because they are not directly observable from the systems inputs and outputs.</p>
</blockquote>
<p>&mdash; Page 31, <a href="https://amzn.to/2vhyW8j">Neural Smithing: Supervised Learning in Feedforward Artificial Neural Networks</a>, 1999.</p>
<p>We can summarize the types of layers in an MLP as follows:</p>
<ul>
<li>
<strong>Input Layer</strong>: Input variables, sometimes called the visible layer.</li>
<li>
<strong>Hidden Layers</strong>: Layers of nodes between the input and output layers. There may be one or more of these layers.</li>
<li>
<strong>Output Layer</strong>: A layer of nodes that produce the output variables.</li>
</ul>
<p>Finally, there are terms used to describe the shape and capability of a neural network; for example:</p>
<ul>
<li>
<strong>Size</strong>: The number of nodes in the model.</li>
<li>
<strong>Width</strong>: The number of nodes in a specific layer.</li>
<li>
<strong>Depth</strong>: The number of layers in a neural network.</li>
<li>
<strong>Capacity</strong>: The type or structure of functions that can be learned by a network configuration. Sometimes called &ldquo;<em>representational capacity</em>&ldquo;.</li>
<li>
<strong>Architecture</strong>: The specific arrangement of the layers and nodes in the network.</li>
</ul>
<h2>How to Count Layers?</h2>
<p>Traditionally, there is some disagreement about how to count the number of layers.</p>
<p>The disagreement centers around whether or not the input layer is counted. There is an argument to suggest it should not be counted because the inputs are not active; they are simply the input variables. We will use this convention; this is also the convention recommended in the book &ldquo;<em>Neural Smithing</em>&ldquo;.</p>
<p>Therefore, an MLP that has an input layer, one hidden layer, and one output layer is a 2-layer MLP.</p>
<p>The structure of an MLP can be summarized using a simple notation.</p>
<p>This convenient notation summarizes both the number of layers and the number of nodes in each layer. The number of nodes in each layer is specified as an integer, in order from the input layer to the output layer, with the size of each layer separated by a forward-slash character (&ldquo;/&rdquo;).</p>
<p>For example, a network with two variables in the input layer, one hidden layer with eight nodes, and an output layer with one node would be described using the notation: 2/8/1.</p>
<p>I recommend using this notation when describing the layers and their size for a Multilayer Perceptron neural network.</p>
<h2>Why Have Multiple Layers?</h2>
<p>Before we look at how many layers to specify, it is important to think about why we would want to have multiple layers.</p>
<p>A single-layer neural network can only be used to represent linearly separable functions. This means very simple problems where, say, the two classes in a classification problem can be neatly separated by a line. If your problem is relatively simple, perhaps a single layer network would be sufficient.</p>
<p>Most problems that we are interested in solving are not linearly separable.</p>
<p>A Multilayer Perceptron can be used to represent convex regions. This means that in effect, they can learn to draw shapes around examples in some high-dimensional space that can separate and classify them, overcoming the limitation of linear separability.</p>
<p>In fact, there is a theoretical finding by Lippmann in the 1987 paper &ldquo;<a href="https://ieeexplore.ieee.org/abstract/document/1165576/">An introduction to computing with neural nets</a>&rdquo; that shows that an MLP with two hidden layers is sufficient for creating classification regions of any desired shape. This is instructive, although it should be noted that no indication of how many nodes to use in each layer or how to learn the weights is given.</p>
<p>A further theoretical finding and proof has shown that MLPs are <a href="https://en.wikipedia.org/wiki/Universal_approximation_theorem">universal approximators</a>. That with one hidden layer, an MLP can approximate any function that we require.</p>
<blockquote>
<p>Specifically, the universal approximation theorem states that a feedforward network with a linear output layer and at least one hidden layer with any &ldquo;squashing&rdquo; activation function (such as the logistic sigmoid activation function) can approximate any Borel measurable function from one finite-dimensional space to another with any desired non-zero amount of error, provided that the network is given enough hidden units.</p>
</blockquote>
<p>&mdash; Page 198, <a href="https://amzn.to/2IXzUIY">Deep Learning</a>, 2016.</p>
<p>This is an often-cited theoretical finding and there is a ton of literature on it. In practice, we again have no idea how many nodes to use in the single hidden layer for a given problem nor how to learn or set their weights effectively. Further, many counterexamples have been presented of functions that cannot directly be learned via a single one-hidden-layer MLP or require an infinite number of nodes.</p>
<p>Even for those functions that can be learned via a sufficiently large one-hidden-layer MLP, it can be more efficient to learn it with two (or more) hidden layers.</p>
<blockquote>
<p>Since a single sufficiently large hidden layer is adequate for approximation of most functions, why would anyone ever use more? One reason hangs on the words &ldquo;sufficiently large&rdquo;. Although a single hidden layer is optimal for some functions, there are others for which a single-hidden-layer-solution is very inefficient compared to solutions with more layers.</p>
</blockquote>
<p>&mdash; Page 38, <a href="https://amzn.to/2vhyW8j">Neural Smithing: Supervised Learning in Feedforward Artificial Neural Networks</a>, 1999.</p>
<h2>How Many Layers and Nodes to Use?</h2>
<p>With the preamble of MLPs out of the way, let&rsquo;s get down to your real question.</p>
<p>How many layers should you use in your Multilayer Perceptron and how many nodes per layer?</p>
<p>In this section, we will enumerate five approaches to solving this problem.</p>
<h3>1) Experimentation</h3>
<p>In general, when I&rsquo;m asked how many layers and nodes to use for an MLP, I often reply:</p>
<blockquote>
<p>I don&rsquo;t know. Use systematic experimentation to discover what works best for your specific dataset.</p>
</blockquote>
<p>I still stand by this answer.</p>
<p>In general, you cannot analytically calculate the number of layers or the number of nodes to use per layer in an artificial neural network to address a specific real-world predictive modeling problem.</p>
<p>The number of layers and the number of nodes in each layer are model hyperparameters that you must specify.</p>
<p>You are likely to be the first person to attempt to address your specific problem with a neural network. No one has solved it before you. Therefore, no one can tell you the answer of how to configure the network.</p>
<p>You must discover the answer using a robust test harness and controlled experiments. For example, see the post:</p>
<ul>
<li><a href="https://machinelearningmastery.com/evaluate-skill-deep-learning-models/">How to Evaluate the Skill of Deep Learning Models</a></li>
</ul>
<p>Regardless of the heuristics you might encounter, all answers will come back to the need for careful experimentation to see what works best for your specific dataset.</p>
<h3>2) Intuition</h3>
<p>The network can be configured via intuition.</p>
<p>For example, you may have an intuition that a deep network is required to address a specific predictive modeling problem.</p>
<p>A deep model provides a hierarchy of layers that build up increasing levels of abstraction from the space of the input variables to the output variables.</p>
<p>Given an understanding of the problem domain, we may believe that a deep hierarchical model is required to sufficiently solve the prediction problem. In which case, we may choose a network configuration that has many layers of depth.</p>
<blockquote>
<p>Choosing a deep model encodes a very general belief that the function we want to learn should involve composition of several simpler functions. This can be interpreted from a representation learning point of view as saying that we believe the learning problem consists of discovering a set of underlying factors of variation that can in turn be described in terms of other, simpler underlying factors of variation.</p>
</blockquote>
<p>&mdash; Page 201, <a href="https://amzn.to/2IXzUIY">Deep Learning</a>, 2016.</p>
<p>This intuition can come from experience with the domain, experience with modeling problems with neural networks, or some mixture of the two.</p>
<p>In my experience, intuitions are often invalidated via experiments.</p>
<h3>3) Go For Depth</h3>
<p>In their important textbook on deep learning, Goodfellow, Bengio, and Courville highlight that empirically, on problems of interest, deep neural networks appear to perform better.</p>
<p>Specifically, they state the choice of using deep neural networks as a statistical argument in cases where depth may be intuitively beneficial.</p>
<blockquote>
<p>Empirically, greater depth does seem to result in better generalization for a wide variety of tasks. [&hellip;] This suggests that using deep architectures does indeed express a useful prior over the space of functions the model learns.</p>
</blockquote>
<p>&mdash; Page 201, <a href="https://amzn.to/2IXzUIY">Deep Learning</a>, 2016.</p>
<p>We may use this argument to suggest that using deep networks, those with many layers, may be a heuristic approach to configuring networks for challenging predictive modeling problems.</p>
<p>This is similar to the advice for starting with <a href="https://machinelearningmastery.com/use-random-forest-testing-179-classifiers-121-datasets/">Random Forest</a> and <a href="https://machinelearningmastery.com/start-with-gradient-boosting/">Stochastic Gradient Boosting</a> on a predictive modeling problem with tabular data to quickly get an idea of an upper-bound on model skill prior to testing other methods.</p>
<h3>4) Borrow Ideas</h3>
<p>A simple, but perhaps time consuming approach, is to leverage findings reported in the literature.</p>
<p>Find research papers that describe the use of MLPs on instances of prediction problems similar in some way to your problem. Note the configuration of the networks used in those papers and use them as a starting point for the configurations to test on your problem.</p>
<p>Transferability of model hyperparameters that result in skillful models from one problem to another is a challenging open problem and the reason why model hyperparameter configuration is more art than science.</p>
<p>Nevertheless, the network layers and number of nodes used on related problems is a good starting point for testing ideas.</p>
<h3>5) Search</h3>
<p>Design an automated search to test different network configurations.</p>
<p>You can seed the search with ideas from literature and intuition.</p>
<p>Some popular search strategies include:</p>
<ul>
<li>
<strong>Random</strong>: Try random configurations of layers and nodes per layer.</li>
<li>
<strong>Grid</strong>: Try a systematic search across the number of layers and nodes per layer.</li>
<li>
<strong>Heuristic</strong>: Try a directed search across configurations such as a genetic algorithm or Bayesian optimization.</li>
<li>
<strong>Exhaustive</strong>: Try all combinations of layers and the number of nodes; it might be feasible for small networks and datasets.</li>
</ul>
<p>This can be challenging with large models, large datasets and combinations of the two. Some ideas to reduce or manage the computational burden include:</p>
<ul>
<li>Fit models on a smaller subset of the training dataset to speed up the search.</li>
<li>Aggressively bound the size of the search space.</li>
<li>Parallelize the search across multiple server instances (e.g. use <a href="https://machinelearningmastery.com/develop-evaluate-large-deep-learning-models-keras-amazon-web-services/">Amazon EC2 service</a>).</li>
</ul>
<p>I recommend being systematic if time and resources permit.</p>
<h3>More</h3>
<p>I have seen countless heuristics of how to estimate the number of layers and either the total number of neurons or the number of neurons per layer.</p>
<p>I do not want to enumerate them; I&rsquo;m skeptical that they add practical value beyond the special cases on which they are demonstrated.</p>
<p>If this area is interesting to you, perhaps start with &ldquo;<em>Section 4.4 Capacity versus Size</em>&rdquo; in the book &ldquo;<a href="https://amzn.to/2vhyW8j">Neural Smithing</a>&ldquo;. It summarizes a ton of findings in this area. The book is dated from 1999, so there are another nearly 20 years of ideas to wade through in this area if you&rsquo;re up for it.</p>
<p>Also, see some of the discussions linked in the <em>Further Reading</em> section (below).</p>
<p>Did I miss your favorite method for configuring a neural network? Or do you know a good reference on the topic?<br>
Let me know in the comments below.</p>
<h2>Further Reading</h2>
<p>This section provides more resources on the topic if you are looking to go deeper.</p>
<h3>Papers</h3>
<ul>
<li>
<a href="https://ieeexplore.ieee.org/abstract/document/1165576/">An introduction to computing with neural nets</a>, 1987.</li>
<li>
<a href="https://www.tandfonline.com/doi/abs/10.1080/01431160802549278">How many hidden layers and nodes?</a>, 2009.</li>
</ul>
<h3>Books</h3>
<ul>
<li>
<a href="https://amzn.to/2vhyW8j">Neural Smithing: Supervised Learning in Feedforward Artificial Neural Networks</a>, 1999.</li>
<li>
<a href="https://amzn.to/2IXzUIY">Deep Learning</a>, 2016.</li>
</ul>
<h3>Articles</h3>
<ul>
<li><a href="https://en.wikipedia.org/wiki/Artificial_neural_network">Artificial neural network on Wikipedia</a></li>
<li><a href="https://en.wikipedia.org/wiki/Universal_approximation_theorem">Universal approximation theorem on Wikipedia</a></li>
<li><a href="http://www.faqs.org/faqs/ai-faq/neural-nets/part3/section-9.html">How many hidden layers should I use?, comp.ai.neural-nets FAQ</a></li>
</ul>
<h3>Discussions</h3>
<ul>
<li><a href="https://stats.stackexchange.com/questions/181/how-to-choose-the-number-of-hidden-layers-and-nodes-in-a-feedforward-neural-netw">How to choose the number of hidden layers and nodes in a feedforward neural network?</a></li>
<li><a href="https://stats.stackexchange.com/questions/160887/number-of-nodes-in-hidden-layers-of-neural-network">Number of nodes in hidden layers of neural network</a></li>
<li><a href="https://stackoverflow.com/questions/10565868/multi-layer-perceptron-mlp-architecture-criteria-for-choosing-number-of-hidde">multi-layer perceptron (MLP) architecture: criteria for choosing number of hidden layers and size of the hidden layer?</a></li>
<li><a href="https://www.quora.com/In-deep-learning-how-do-I-select-the-optimal-number-of-layers-and-neurons">In deep learning, how do I select the optimal number of layers and neurons?</a></li>
</ul>
<h2>Summary</h2>
<p>In this post, you discovered the role of layers and nodes and how to configure a multilayer perceptron neural network.</p>
<p>Specifically, you learned:</p>
<ul>
<li>The difference between single-layer and multiple-layer perceptron networks.</li>
<li>The value of having one and more than one hidden layers in a network.</li>
<li>Five approaches for configuring the number of layers and nodes in a network.</li>
</ul>
<p>Do you have any questions?<br>
Ask your questions in the comments below and I will do my best to answer.</p>
<p>The post <a rel="nofollow" href="https://machinelearningmastery.com/how-to-configure-the-number-of-layers-and-nodes-in-a-neural-network/">How to Configure the Number of Layers and Nodes in a Neural Network</a> appeared first on <a rel="nofollow" href="https://machinelearningmastery.com">Machine Learning Mastery</a>.</p>
<br>
<br>
<br>
<br>
AI|ML|DM|IR|PR<br>
<br>
via Machine Learning Mastery https://ift.tt/2if3vo8<br>
<br>
July 27, 2018 at 03:17AM
